authz/providers/azure/rbac/rbac.go

```diff
@@ -0,0 +1,198 @@
+		return errors.Errorf("%s: failed to refresh token: %s", a.tokenProvider.Name(), err)
+	}
+
+	// Set the authorization headers for future requests
+	a.headers.Set("Authorization", fmt.Sprintf("Bearer %s", resp.Token))
+	expIn := time.Duration(resp.Expires) * time.Second
+	a.expires = time.Now().Add(expIn - expiryDelta)
+
+	return nil
+}
+
+func (a *AccessInfo) IsTokenExpired() bool {
+	if a.expires.Before(time.Now()) {
+		return true
+	} else {
+		return false
+	}
+}
+
+func (a *AccessInfo) CheckAccess(request *authzv1.SubjectAccessReviewSpec) (*authzv1.SubjectAccessReviewStatus, error) {
+	var CHECKACCESS_PATH string = "/providers/Microsoft.Authorization/checkaccess"
+	var API_VERSION string = "2018-09-01-preview"
+	checkAccessBody := PrepareCheckAccessRequest(request, a.clusterType, a.azureResourceId)
+	checkAccessURL := *a.apiURL
+	// Append the path for azure cluster resource id
+	checkAccessURL.Path = path.Join(checkAccessURL.Path, a.azureResourceId)
+	var str string
+	if getNameSpaceScope(request, &str) {
+		checkAccessURL.Path = path.Join(checkAccessURL.Path, str)
+	}
+
+	checkAccessURL.Path = path.Join(checkAccessURL.Path, CHECKACCESS_PATH)
+	params := url.Values{}
+	params.Add("api-version", API_VERSION)
+	checkAccessURL.RawQuery = params.Encode()
+
+	if a.IsTokenExpired() {
+		a.RefreshToken()
+	}
+
+	buf := new(bytes.Buffer)
+	if err := json.NewEncoder(buf).Encode(checkAccessBody); err != nil {
+		return nil, errors.Wrap(err, "error encoding check access request")
+	}
+
+	binaryData, _ := json.MarshalIndent(checkAccessBody, "", "    ")
+	glog.Infof("checkAccessURI:%s", checkAccessURL.String())
+	glog.Infof("binary data:%s", binaryData)
+
+	req, err := http.NewRequest(http.MethodPost, checkAccessURL.String(), buf)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating check access request")
+	}
+	// Set the auth headers for the request
+	req.Header = a.headers
+
+	if glog.V(10) {
+		cmd, _ := http2curl.GetCurlCommand(req)
+		glog.V(10).Infoln(cmd)
+	}
+
+	resp, err := a.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "error in check access request execution")
+	}
+	defer resp.Body.Close()
+
+	data, _ := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != http.StatusOK {
+		glog.Errorf("error in check access response. error code: %d, response: %s", string(binaryData))
+		if resp.StatusCode == http.StatusTooManyRequests {
+			glog.V(10).Infoln("Moving to another ARM instance!")
+			a.client.CloseIdleConnections()
+			//to-do retry for this
+			// add metrix for this scenario
+			return &authzv1.SubjectAccessReviewStatus{Allowed: false, Reason: "server error", Denied: true},
+				errors.Errorf("request %s failed with status code: %d and response: %s", req.URL.Path, resp.StatusCode, string(data))
+		}
+
+		if resp.StatusCode >= http.StatusInternalServerError {
+			return &authzv1.SubjectAccessReviewStatus{Allowed: false, Reason: "server error", Denied: false},
```

Comment:
Denied: false?

Reply:
This was before our discussion. Will update as per closure.